### PR TITLE
Rename `wasm32-wasi` to `wasm32-wasi-preview1`

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Implements https://github.com/rust-lang/compiler-team/issues/607

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
Marking as waiting on author, given the WIP state; feel free to ping me/change labels when ready.

**Comment by yoshuawuyts** [comment]
This compiles locally now, and CI seems to be passing!

@WaffleLapkin can I ask you to maybe take a look now?

File: config.example.toml
Comment (WaffleLapkin) [inline config.example.toml:769]: This is not right, to many previews.

**Review by WaffleLapkin — COMMENTED** [review]
I think the MCP mentioned the period where the old name would be accepted with the new name, but it looks like this PR just renames the target, without handling the old name in any way, why so?

Additionally I've found a few mentions of `wasm32-wasi` in the repo, that were not changed:
```text
library/stdarch/ci/docker/wasm32-wasi/Dockerfile
15:  --mapdir .::/checkout/target/wasm32-wasi/release/deps

src/stage0.json
224:    "dist/2023-03-07/rust-std-beta-wasm32-wasi.tar.gz": "265fa8b315a5d39a35fb8d32d5e46c3c66f9608992a3d708ac90437818cfed45",
225:    "dist/2023-03-07/rust-std-beta-wasm32-wasi.tar.xz": "26839f3ed020dbda8ba893492cf504d565e7e1af7cfec5ad76053443d1022839",

src/doc/rustc-dev-guide/src/building/how-to-build-and-run.md
221:and your cross-compilation target is `wasm32-wasi`, you can build with:
224:./x.py build --target x86_64-unknown-linux-gnu --target wasm32-wasi
236:target = ["x86_64-unknown-linux-gnu", "wasm32-wasi"]
257:cargo +stage1 build --target wasm32-wasi
```
Can you clarify why these are not/should not be changed?

File: compiler/rustc_target/src/spec/wasm32_wasi_preview1.rs
Comment (yoshuawuyts) [reply compiler/rustc_target/src/spec/wasm32_wasi_preview1.rs:59]: > Is `-preview1` not supported by other tooling?

Yes, that's right. The rename has been recent, and not everyone has adopted it yet.

**Comment by yoshuawuyts** [comment]
> I think the MCP mentioned the period where the old name would be accepted with the new name, but it looks like this PR just renames the target, without handling the old name in any way, why so?

@WaffleLapkin This is something we explicitly decided not to do. I think perhaps you might be confusing it with both Preview 1 and Preview 2 existing in the compiler for the foreseeable future? Relevant text from the MCP:

> Because we are not (yet) removing a target but only renaming it, we should be able to do this over a regular Rust release. This means we are **not** going to have a period where the `wasm32-wasi` and `wasm32-wasi-preview1` targets are both available in the toolchain.

---

> Additionally I've found a few mentions of wasm32-wasi in the repo, that were not changed:

Oops! Thank you for spotting that! - The first and third one are reason other than I don't remember it showing up when I grepped the codebase.

The second one seems to be describing some sort of mapping of hashes and targets though. I actually don't know what that is, or how I should update that? Do I just rename the target, and will the hash remain the same? I could just try it and see what happens?

File: config.example.toml
Comment (yoshuawuyts) [reply config.example.toml:769]: oops, good catch. Fixed!

**Comment by yoshuawuyts** [comment]
>  I'll ask the bootstrap team on Zulip what to do with this.

I asked the bootstrap team, and they told me not to touch `src/stage0.json`. It will get auto-regenerated when the next beta release is cut - so we don't need to do anything with that right now.

**Comment by yoshuawuyts** [comment]
Okay, I've tracked down the other two missing renames of `wasm32-wasi`, and I need to update different repos for that. Namely:

- [rustc-dev-guide](https://github.com/rust-lang/rustc-dev-guide)
- [rust-lang/stdarch](https://github.com/rust-lang/stdarch)

I'll file separate patches to both those repos. And I think with that all feedback from review has been handled.

File: config.example.toml
Comment (WaffleLapkin) [inline config.example.toml:769]: This still says `preview1-preview1`

**Review by WaffleLapkin — APPROVED** [review]
r=me with the last nitpick fixed

**Comment by yoshuawuyts** [comment]
@WaffleLapkin thanks for catching that; fixed!

**Comment by klensy** [comment]
I understand correct, this *renames* `wasm32-wasi` to `wasm32-wasi-preview1`, without adding alias between them (at least temporary, to prevent breakage) and intentionally breaking documentation all over internet.

**Comment by wesleywiser** [comment]
@yoshuawuyts Can you write a blog for the Rust blog announcing this change? It would be great if we can coordinate those together with this.

**Comment by codefromthecrypt** [comment]
fyi go shortened this to "wasip1"

**Comment by yoshuawuyts** [comment]
@wesleywiser https://github.com/rust-lang/blog.rust-lang.org/pull/1099 is now out for review. If that looks good, I think this can be merged?

**Comment by jyn514** [comment]
Let's please do a dev-static test before merging this: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/unavailable.20toolchain.20upgrade.20behavior.3F. @Mark-Simulacrum has permissions to kick off a dev-static build and I think @pietroalbini might as well, but I don't.

There was also an unresolved concern about whether this should have an alias or not; if we add an alias for at least a few weeks I think it will make the transition smoother since `rustup update` won't give a hard error: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/unavailable.20toolchain.20upgrade.20behavior.3F/near/352551411

**Comment by yoshuawuyts** [comment]
> Let's please do a dev-static test before merging this […]

I don't quite understand what a "dev-static test" is, but if you think this is a good idea I don't have any objections.

---

> There was also an unresolved concern about whether this should have an alias or not […]

That depends on what you mean by "alias" here. [The MCP](https://github.com/rust-lang/compiler-team/issues/607) seems pretty clear that we don't want to allow both `wasm32-wasi` and `wasm32-wasi-preview1` to be simultaneously available. We decided that doing this over a single release would be the best way of making this change, keeping the transition short and sweet. What we're doing here is a target rename. This means deprecating an existing target, and bringing it back under a different name. That's fundamentally a breaking change, and we know will surface to end-users.

The overriding concern here should be that we clearly communicate the target rename. When people's workflows break, they're going to want instructions on how to unbreak it. The main vehicle for this will be the main Rust blog. I think it's a good idea to further explore ways in which we can smooth the transition. But I want to make sure that we treat those as *further improvements*, and not as things that warrant blocking on.

So I'd like us to see the current set of PRs through first. Once they've landed, we'll be able to observe how the change went over on nightly. From talking with folks offline, the expectation is that this will go roughly the same as when the NDK versions are bumped. If we unexpectedly run into any communication issues, this gives us more than enough time to address those before the change goes to stable.

**Comment by CryZe** [comment]
I'm not entirely convinced that this is the right thing to do. I build a little ecosystem where increasingly more people are starting to write little projects for with Rust + WASM. A lot of those people are actually very inexperienced in Rust (and probably WASM too). So they are copying projects from each other to start their projects. They will never read the Rust blog post about the old target being removed, as they aren't even aware of that blog in the first place. Okay, but let's say they all make the switch to the new target. Well people will still keep copying these projects to bootstrap theirs, so there's even more people who are unaware of all of this afterwards, but here's the thing: Eventually the preview1 target is going to be removed, as WASI matures. That means a lot of their build processes will fail, yet again (and most of these people aren't even aware of the first time it broke). But the thing is, they never actually cared about any specific WASI version in the first place, because they are all targeting the most recent versions of wasmtime. So not having a way to express "just target the latest WASI" seems like a huge amount of problems that this is going to introduce for the people in my ecosystem with literally zero benefit.

**Comment by yoshuawuyts** [comment]
@CryZe I'm sorry you have concerns; but can we please not re-litigate the contents of the MCP in this PR? This really isn't the right venue for that.

**Comment by CryZe** [comment]
Do I open an MCP or what is the right venue to discuss that?

**Comment by jyn514** [comment]
You can comment in the existing MCP thread. If a t-compiler member agrees with you, we can discuss what to do from there. However, in general I am fairly opposed to changing strategy after an MCP has already been accepted unless there's significant new information.

**Comment by yoshuawuyts** [comment]
This PR currently seems stuck on potentially needing to run a dev-static build (see: https://github.com/rust-lang/rust/pull/110596#issuecomment-1523805921)

r? @Mark-Simulacrum

**Comment by jyn514** [comment]
FYI Mark only looks at PRs marked waiting-on-review.

@rustbot ready

**Comment by yoshuawuyts** [comment]
ping @rust-lang/compiler - This PR has been sitting idle for almost a month now. I would like to request we do one of two things here:

1. Decide we actually do need to run a "dev-static" test and then run one. This needs someone to step forward to run it, and in the case of a failure help navigate the process to get it to pass. Currently @Mark-Simulacrum is assigned.
2. Decide the benefits of running a "dev-static" test are not worth it at this time, and proceed to merge this PR.

Right now it feels like this PR has hit a limbo state where it's unclear what to do next / who should take the next steps. I would very much like to see this PR land, so I'm requesting we take either action to move this forward. Thank you!

**Comment by Mark-Simulacrum** [comment]
I expect to push and run that test in the next week, just hasn't been in my queue until 4 days ago.

**Comment by wesleywiser** [comment]
> > What we're doing here is a target rename. This means deprecating an existing target, and bringing it back under a different name. That's fundamentally a breaking change, and we know will surface to end-users.
> 
> Right. There's two parts of a target, the name you can pass to `--target` and the name you can pass to `rustup target add`. @bjorn3's suggest was to add an alias for the latter but not the former.

I think this is a good suggestion! If this is paired with a more helpful error from rustc about the target name change, that would also be great:

```
$ rustup target add --toolchain stable wasm32-wasi
info: downloading component 'rust-std' for 'wasm32-wasi-preview1'
info: installing component 'rust-std' for 'wasm32-wasi-preview1'
 18.9 MiB /  18.9 MiB (100 %)  15.9 MiB/s in  1s ETA:  0s

$ cargo build --target wasm32-wasi
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `rustc - --crate-name ___ --print=file-names --target wasm32-wasi --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 1)
  --- stderr
  error: Error loading target specification: Could not find specification for target "wasm32-wasi". Run `rustc --print target-list` for a list of built-in targets
  help: "wasm32-wasi" has been renamed to "wasm32-wasi-preview1"

$ cargo build --target wasm32-wasi-preview1  # rustup already installed the right target so this will work
   Compiling hello_world v0.1.0 (/home/wesley/hello_world)
    Finished dev [unoptimized + debuginfo] target(s) in 0.41s
```

I don't know if rustup will mention the rename to the user but that would be even better.

**Comment by Mark-Simulacrum** [comment]
@bors try

Getting artifacts with this PRs changes to generate a dev-static nightly release...

**Comment by Mark-Simulacrum** [comment]
@yoshuawuyts I think this may need some changes on your end -- sorry for the delay here. Once we get a try build working the dev-static bit shouldn't take long.

**Comment by yoshuawuyts** [comment]
I am very confused by the failure. This seems like it should just be an easy string matching test, and it seems to work without a problem locally. But for some reason it's failing on CI - the test runner is just linux, so there should be nothing special about the environment either. The command to run the tests locally is:

```bash
./x test src/tools/compiletest --stage 2
```

**Comment by jyn514** [comment]
@yoshuawuyts i don't know why it doesn't fail locally, but it looks like it's indexing into a map of targets - it makes sense to me you'd need to update the target name: https://github.com/rust-lang/rust/blob/4dbc7e3092fce6f3d5712733e5206226183dbe1e/src/tools/compiletest/src/common.rs#L499
https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/header/tests.rs#L449-L453

**Comment by yoshuawuyts** [comment]
Okay, interesting. I can get the test failure to repro locally if I change the `wasm_special` test back to `wasm32-wasi` from `wasm32-wasi-preview1`:

```rust
        ("wasm32-wasi", "emscripten", false),
        ("wasm32-wasi", "wasm32", true),
        ("wasm32-wasi", "wasm32-bare", false),
        ("wasm32-wasi", "wasi", true),
```

This to me indicates that somehow we're generating some something locally with the new names which isn't being generated on CI. The likely fix seems to be to make sure we update what's being generated?

**Comment by yoshuawuyts** [comment]
Found the issue with the help of @jyn514: mingw runs on stage 0 which doesn't know about the target renames. We can repro it locally by running:
```bash
./x test src/tools/compiletest --stage 0
```

The solution we're discussing is to ensure compiletest is aware of stages, and temporarily make the wasm test conditional just for stage 0.

**Comment by jyn514** [comment]
we figured this out, mingw-check uses stage 0 so compiletest loads the target list from the beta compiler

File: src/tools/compiletest/src/header/tests.rs
Comment (jyn514) [inline src/tools/compiletest/src/header/tests.rs:458]:
```suggestion
    // removed, and the `else` part should be re-inlined in the original `ignores` list.
    // cfg(bootstrap)
```
this tells the person updating the beta compiler that this should be removed

File: src/bootstrap/test.rs
Comment (jyn514) [inline src/bootstrap/test.rs:707]: i don't think this change is necessary — i'd prefer to keep using builder.top_stage in run().

**Review by jyn514 — COMMENTED** [review]
thanks, the compiletest change looks roughly like what i expected :) nice work!

**Comment by yoshuawuyts** [comment]
I've updated the PR with feedback from review, and also created a tracking issue in https://github.com/rust-lang/rust/issues/113364 to keep track of which other changes need to be made. In order to achieve the flow proposed in https://github.com/rust-lang/rust/pull/110596#issuecomment-1571012470 we'll need to make subsequent patches to rustup and cargo which I'll be working on next.

**Comment by jyn514** [comment]
@Mark-Simulacrum i think this should be working now - could you kick off a dev-static release please?

**Comment by Mark-Simulacrum** [comment]
Kicked off a dev-static [release](https://forge.rust-lang.org/release/process.html#publishing-a-nightly-based-off-a-try-build). Should be ready in 1-1.5 hours.

**Comment by Mark-Simulacrum** [comment]
The dev-static build failed due to missing docs, I think due to #111660 dropping docs from try builds. However, I think this PR didn't end up enabling wasm32-wasi anyway?

I pushed commits that should fix that, let's run another try build. @bors try

**Comment by g0djan** [comment]
Hi here, I'm working on adding `wasm32-wasi-threads` https://github.com/rust-lang/rust/pull/112922 target and @deltragon pointed out that there might be not much sense to name it like that cause it needs to be renamed right away.

Should I give it a name includeing `preview1`? It seems there might be 2 options:
- `wasm32-wasi-threads-preview1`
- `wasm32-wasi-preview1-threads`

Cc: @abrown @bjorn3 @alexcrichton

**Comment by yoshuawuyts** [comment]
@g0djan The name for that target we proposed in https://github.com/rust-lang/compiler-team/issues/607 was `wasm32-wasi-preview1-threads` - since `-threads` is considered an extension to the base `wasm32-wasi-preview1` target.

**Comment by codefromthecrypt** [comment]
why would there be a specific wasm32-wasi-preview1-threads? since threads isn't related to preview1? is it to avoid a common compiler flag?

**Comment by yoshuawuyts** [comment]
@codefromthecrypt this is getting a little off-topic here, but if you want to read more you can find the accepted compiler MCP for it here: https://github.com/rust-lang/compiler-team/issues/574

**Comment by tshepang** [comment]
am thinking the name should actually be `wasm32-wasi-preview2-threads`, for it is based on `preview2`

**Comment by codefromthecrypt** [comment]
unsolicited 2p

the main thing is lets try not to make an already confusing ecosystem (wasm) more confused by conflating things that are incompatible or not really extensions. Features that aren't finished should probably not be formalized as a triple and if they are, coherently somehow. This feedback would be wildcard applied to any similar topic, if rust ends up with a cartesian product of features in triples somehow (exaggerating to make a point I know you won't). I would encourage asking diverse folks to comment before merging new triples, as I think perspective helps. It is completely your call, and if something is merged, you can of course you can consider it immutable. OTOH I would recommend in something that's likely to be contested trying to find feedback vs relying on watchers at the time.

**Comment by g0djan** [comment]
> this is getting a little off-topic here

I moved the discussion to Zulip https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/New.20tier-2.20target.20for.20.60wasm32-wasi.60.20.2B.20thr.E2.80.A6.20compiler-team.23574/near/374571151

**Comment by wesleywiser** [comment]
(it's still marked as "trying" in the bors queue even after @workingjubilee closed and reopened. Trying again...)

**Comment by wesleywiser** [comment]
Wait... now it's "pending" not "pending (try)"

@bors r-

**Comment by wesleywiser** [comment]
Hi @Mark-Simulacrum, we've gotten the try build sorted out and completed. Could you kick off a dev-static release again? Thanks!

File: src/bootstrap/compile.rs
Comment (g0djan) [inline src/bootstrap/compile.rs:309]: Is it able to find wasi-libc sysroot? My understanding wasi-libc targets doesn't mention `-preview1`

Comment (g0djan) [reply src/bootstrap/compile.rs:309]: Also I might be doing something wrong though, but could you try to run it yourself please @yoshuawuyts

Comment (g0djan) [reply src/bootstrap/compile.rs:309]: tried to fix it by removing `-preview1` form here, but it doesn't fix the problem

Comment (tomaka) [reply src/bootstrap/compile.rs:309]: > tried to fix it by removing -preview1 from here, but it doesn't fix the problem

You most likely have to revert the other change in the same file (line 411) as well.

Comment (g0djan) [reply src/bootstrap/compile.rs:309]: Yeah I did, stiil the same

**Comment by Mark-Simulacrum** [comment]
Apologies for the delay, I've kicked off dev-static nightly for e1e8b3f26b4661e6bcda6a18408e8fc8b5ed2d8a now.

**Comment by jeffparsons** [comment]
There's a wasi-libc PR up for renaming `wasm32-wasi-threads` to `wasm32-wasi-preview1-threads`: https://github.com/WebAssembly/wasi-libc/pull/434.

I don't think it touches the `wasm32-wasi` target, though. Maybe best to let that PR land and then put up an equivalent PR?

**Comment by Dylan-DPC** [comment]
@yoshuawuyts what's the update on this?

**Comment by alexcrichton** [comment]
I've opted to (read, my memory is so bad I forgot about this, again) make a "fresh" version of this at https://github.com/rust-lang/rust/pull/120468 to implement the first step of the transition plan in https://github.com/rust-lang/compiler-team/issues/695

**Comment by yoshuawuyts** [comment]
Closing this in favor of https://github.com/rust-lang/compiler-team/issues/695

**Pull request closed without merging**
